runtime/common/src/precompile/liquid_crowdloan.rs

```diff
@@ -0,0 +1,286 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use super::{
+	input::{Input, InputPricer, InputT, Output},
+	target_gas_limit,
+};
+use crate::WeightToGas;
+use frame_support::log;
+use module_evm::{
+	precompiles::Precompile,
+	runner::state::{PrecompileFailure, PrecompileOutput, PrecompileResult},
+	Context, ExitError, ExitRevert, ExitSucceed,
+};
+use module_liquid_crowdloan::WeightInfo;
+use module_support::Erc20InfoMapping as _;
+use num_enum::{IntoPrimitive, TryFromPrimitive};
+use sp_core::Get;
+use sp_runtime::{traits::Convert, RuntimeDebug};
+use sp_std::{marker::PhantomData, prelude::*};
+
+/// The `LiquidCrowdloan` impl precompile.
+pub struct LiquidCrowdloanPrecompile<R>(PhantomData<R>);
+
+#[module_evm_utility_macro::generate_function_selector]
+#[derive(RuntimeDebug, Eq, PartialEq, TryFromPrimitive, IntoPrimitive)]
+#[repr(u32)]
+pub enum Action {
+	Redeem = "redeem(address,uint256)",
+	GetRedeemCurrency = "getRedeemCurrency()",
+}
+
+impl<Runtime> Precompile for LiquidCrowdloanPrecompile<Runtime>
+where
+	Runtime: module_evm::Config + module_prices::Config + module_liquid_crowdloan::Config,
+{
+	fn execute(input: &[u8], target_gas: Option<u64>, _context: &Context, _is_static: bool) -> PrecompileResult {
+		let input = Input::<Action, Runtime::AccountId, Runtime::AddressMapping, Runtime::Erc20InfoMapping>::new(
+			input,
+			target_gas_limit(target_gas),
+		);
+
+		let gas_cost = Pricer::<Runtime>::cost(&input)?;
+
+		if let Some(gas_limit) = target_gas {
+			if gas_limit < gas_cost {
+				return Err(PrecompileFailure::Error {
+					exit_status: ExitError::OutOfGas,
+				});
+			}
+		}
+
+		let action = input.action()?;
+
+		match action {
+			Action::Redeem => {
+				let who = input.account_id_at(1)?;
+				let amount = input.balance_at(2)?;
+
+				let redeem_amount =
+					<module_liquid_crowdloan::Pallet<Runtime>>::do_redeem(&who, amount).map_err(|e| {
+						PrecompileFailure::Revert {
+							exit_status: ExitRevert::Reverted,
+							output: Output::encode_error_msg("LiquidCrowdloan redeem failed", e),
+							cost: target_gas_limit(target_gas).unwrap_or_default(),
+						}
+					})?;
+
+				log::debug!(target: "evm", "liuqid_crowdloan: Redeem who: {:?}, amount: {:?}, output: {:?}", who, amount, redeem_amount);
+				Ok(PrecompileOutput {
+					exit_status: ExitSucceed::Returned,
+					cost: gas_cost,
+					output: Output::encode_uint(redeem_amount),
+					logs: Default::default(),
+				})
+			}
+			Action::GetRedeemCurrency => {
+				let currency_id = <module_liquid_crowdloan::Pallet<Runtime>>::redeem_currency();
+				let address = <Runtime as module_prices::Config>::Erc20InfoMapping::encode_evm_address(currency_id)
+					.unwrap_or_default();
```

Comment:
It's better to return `Revert` here

Reply:
This is a read operation. It better to return empty address and let contract handle it

Reply:
in this particular case, it is something should never happen and I think revert is fine

Reply:
just for consistency. every read operation shouldn't revert